src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/Sources/GeneratedOutputSource.cs

```diff
@@ -10,60 +10,65 @@
 
 namespace Microsoft.CodeAnalysis.Razor.ProjectSystem.Sources;
 
-internal sealed class GeneratedOutputSource(DocumentSnapshot document)
+internal sealed class GeneratedOutputSource
 {
-    private readonly DocumentSnapshot _document = document;
     private readonly SemaphoreSlim _gate = new(initialCount: 1);
 
     // Hold the output in a WeakReference to avoid memory leaks in the case of a long-lived
     // document snapshots. In particular, the DynamicFileInfo system results in the Roslyn
     // workspace holding onto document snapshots.
-    private WeakReference<RazorCodeDocument>? _output;
+    private WeakReference<RazorCodeDocument>? _weakOutput;
+
+    public GeneratedOutputSource ForkIfOutputAvailable()
```

Comment:
❓why is it safe to call `TryGetValue` without using the `_gate` here?

Reply:
Why using `_gate` be needed? `TryGetValue` performs a field read of a pointer size, which are guaranteed to be atomic by the runtime. All of the writes are performed under the gate. Note that Roslyn follows the [same](https://source.dot.net/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/VersionSource/LoadableTextAndVersionSource.cs) pattern in many situations.

Reply:
On mobile for a bit so far thumb small screen: I understand the read is optimistic but on line 52 it's under a gate before the write. It always looks odd to have the same code where one is gated and one isn't so wasn't sure if there was a behavior I was unaware of.

Reply:
Oh I got it. Line 41 does the check outside the lock first. So it's the normal 

``` 
If (Get(...))
{
   return...
}

lock(...)
{
    If (Get(...))
    {
          return...
     }
}